zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js

```diff
@@ -1120,6 +1135,201 @@ angular.module('zeppelinWebApp').controller('ParagraphCtrl', function($scope, $r
+      if (pinLayer) {
+        $scope.map.map.remove(pinLayer);
+      }
+
+      // add pins to map as layer
+      getMapPins(function(pins) {
+        createPinMapLayer(pins, function(pinLayer) {
+          $scope.map.map.add(pinLayer);
+          if (pinLayer.source.length > 0) {
+            $scope.map.goTo(pinLayer.source);
+          }
+        });
+      });
+    };
+
+    var createMap = function(mapdiv) {
+      // prevent zooming with the scroll wheel
+      var disableZoom = function(e) {
+        var evt = e || window.event;
+        evt.cancelBubble = true;
+        evt.returnValue = false;
+        if (evt.stopPropagation) {
+          evt.stopPropagation();
+        }
+      };
+      var eName = window.WheelEvent ? 'wheel' :  // Modern browsers
+                  window.MouseWheelEvent ? 'mousewheel' :  // WebKit and IE
+                  'DOMMouseScroll';  // Old Firefox
+      mapdiv.addEventListener(eName, disableZoom, true);
+
+      esriLoader.require(['esri/views/MapView',
+                          'esri/Map',
+                          'esri/renderers/SimpleRenderer',
+                          'esri/symbols/SimpleMarkerSymbol'],
+                          function(MapView, Map, SimpleRenderer, SimpleMarkerSymbol) {
+        $scope.map = new MapView({
+          container: mapdiv,
+          map: new Map({
+            basemap: $scope.paragraph.config.graph.map.baseMapType.toLowerCase()
+          }),
+          center: [-106.3468, 56.1304],  // Canada (lng, lat)
+          zoom: 2,
+          pinRenderer: new SimpleRenderer({
+            symbol: new SimpleMarkerSymbol({
+              'color': [255, 0, 0, 0.5],
+              'size': 16.5,
+              'outline': {
+                'color': [0, 0, 0, 1],
+                'width': 1.125,
+              },
+              // map pin SVG path
+              'path': 'M16,3.5c-4.142,0-7.5,3.358-7.5,7.5c0,4.143,7.5,18.121,7.5,' +
+                      '18.121S23.5,15.143,23.5,11C23.5,6.858,20.143,3.5,16,3.5z ' +
+                      'M16,14.584c-1.979,0-3.584-1.604-3.584-3.584S14.021,7.416,' +
+                      '16,7.416S19.584,9.021,19.584,11S17.979,14.584,16,14.584z'
+            })
+          })
+        });
+
+        $scope.map.on('click', function() {
+          // ArcGIS JS API 4.0 does not account for scrolling or position
+          // changes by default (this is a bug, to be fixed in the upcoming
+          // version 4.1; see https://geonet.esri.com/thread/177238#comment-609681).
+          // This results in a misaligned popup.
+
+          // Workaround: manually set popup position to match position of selected pin
+          if ($scope.map.popup.selectedFeature) {
+            $scope.map.popup.location = $scope.map.popup.selectedFeature.geometry;
+          }
+        });
+        $scope.map.then(updateMapPins);
+      });
+    };
+
+    var renderMap = function() {
+      var mapdiv = angular.element('#p' + $scope.paragraph.id + '_map')
```

Comment:
@corneadoug It is done the way it is currently so that the `#p_ID_map` div can have its height adjusted while also selecting its child div using cascading. I can do it without the cascading however that would require an additional DOM query:

e.g.,
```js
var mapdiv = angular.element('#p' + $scope.paragraph.id + '_map > div').get(0);
angular.element('#p' + $scope.paragraph.id + '_map').css('height', $scope.paragraph.config.graph.height);
```

versus (the current)

```js
var mapdiv = angular.element('#p' + $scope.paragraph.id + '_map')
                    .css('height', $scope.paragraph.config.graph.height)
                    .children('div').get(0);
```